### DISCUSSION TITLE
Are direct exports of Remix actions possible without causing module side effects?

### DISCUSSION THREAD

**Discussion by janhesters** [remix-run/remix#6531] [Q&A] [answered]
In the Remix docs there is a [section on higher-order functions](https://remix.run/docs/en/1.16.1/guides/constraints#higher-order-functions).

They say that something like this:

```ts
export const action = middleware(request => {
  const isJim = new URL(request.url).pathname.includes('jim');
  return json({ name: isJim ? 'Jim' : 'John' });
});
```

doesn't work in Remix because it is a module side effect so the compiler can't prune out the `middleware` code.

I was wondering, if a direct export of a Remix action is possible?

So something like this:

```
import myAction from '~/features/my-action.server';

export const action = myAction;
```

or would the compiler experience the same problem?

**Comment by sergiodxa** [comment, marked as answer]
That should work, because the action comes from a `.server` code, if you use named exports you could even do `export { action } from "~/features/my-action.server"` to do it in one line

**Comment by janhesters** [comment]
@sergiodxa Great! Thank you for your help 🙏 

I'm a huge fan of higher order functions for middleware and it sounds like you're saying if I compose my middleware in a `.server` file, I can import and export them, correct?

```ts
export const teamMembersPageLoader = asyncPipe(
  withOrganizationMembership,
  withLocalization,
  withPageTitle({ tKey: 'team-members:team-members' }),
  withHeaderTitle('team-members:team-members'),
  withTotalMemberCount,
  withCurrentPage,
  withMembersAndLatestInviteLink,
  withTeamMemberPageProps,
  json,
);
```

and then

```ts
export { teamMembersPageLoader as action } from "~/features/team-member-handler.server";
```

should work? The code compiles when I do that at least 🙏

**Reply by sergiodxa** [reply]
Yes, that should work
